### PR TITLE
More bulk persistence

### PR DESCRIPTION
**Pull request by CGDS**
PR implements ordered persistence buffer operations for policy elements and assignments.  If the assignments approach is acceptable, a similar one will be implemented for link operations.  Associations should be done in a following PR, as they arent strictly needed for the MVP we're after, but they will work basically like this too.

### CODE REVIEW

File: lib/policy_machine/policy_element.rb [hunk 1]
```diff
@@ -65,7 +65,9 @@ def unassign(dst_policy_element)
     # Removes self and any assignments to or from self. Does not remove any other elements.
     # Returns true if persisted object was successfully removed.
     def delete
-      if self.stored_pe && self.stored_pe.persisted
+      # If the stored ppe is not persisted, no need to delete it, but if buffering
+      # all bets are off, and the storage adapter needs to handle the situation
+      if (self.stored_pe && self.stored_pe.persisted) || pm_storage_adapter.buffering?
```
Comment (DMcKinnon-mdsol) [inline lib/policy_machine/policy_element.rb]: Should we check if it's buffering first so that we don't have to make that extra db call unless we have to?

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 2]
```diff
@@ -184,11 +187,18 @@ def self.bulk_destroy(elements)
...
 
-      def self.bulk_assign(parents_and_children)
-        id_pairs = parents_and_children.map { |parent, child| [parent.id, child.id]  }
+      def self.bulk_assign(pairs_hash)
+        id_pairs = pairs_hash.values.map { |parent, child| [parent.id, child.id]  }
         Assignment.import([:parent_id, :child_id], id_pairs, on_duplicate_key_ignore: true)
       end
 
+      def self.bulk_unassign(pairs_hash)
+        pairs_str = pairs_hash.values.reduce([]) do |memo,(parent,child)|
```
Comment (DMcKinnon-mdsol) [inline lib/policy_machine_storage_adapters/active_record.rb]: The block args could use some spaces after commas.

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 3]
```diff
@@ -184,11 +187,18 @@ def self.bulk_destroy(elements)
...
+          parent.persisted && child.persisted? ? memo + ["(#{parent.id},#{child.id})"] : memo
```
Comment (DMcKinnon-mdsol) [inline lib/policy_machine_storage_adapters/active_record.rb]: Why use `persisted` and then `persisted`?

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 4]
```diff
@@ -184,11 +187,18 @@ def self.bulk_destroy(elements)
...
+        end.join(',')
```
Comment (DMcKinnon-mdsol) [inline lib/policy_machine_storage_adapters/active_record.rb]: The derivation of `paris_str` scares me. I'm having  a hard time picturing what it'll look like when it's done with this join. Can you provide some examples?

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 5]
```diff
@@ -376,7 +386,8 @@ def assign(src, dst)
     end
 
     def assign_later(parent:, child:)
-      buffers[:assignments] << [parent, child]
+      # buffers[:assignments] << [parent, child]
```
Comment (DMcKinnon-mdsol) [inline lib/policy_machine_storage_adapters/active_record.rb]: Commented code

**Comment by DMcKinnon-mdsol** [comment]
Yea, this method works for me.

File: lib/policy_machine/policy_element.rb (hunk 1 above)
Comment (CGDS) [inline lib/policy_machine/policy_element.rb]: good call, will change

File: lib/policy_machine_storage_adapters/active_record.rb (hunk 2 above)
Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: fiiine

File: lib/policy_machine_storage_adapters/active_record.rb (hunk 3 above)
Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: because im bad at typing.  will fix.

File: lib/policy_machine_storage_adapters/active_record.rb (hunk 4 above)
Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: "(1,2),(3,4)"

Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: or the empty string, if no pairs are persisted

File: lib/policy_machine_storage_adapters/active_record.rb (hunk 5 above)
Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: whoops

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 6]
```diff
@@ -55,8 +55,10 @@ def self.buffers
       @buffers ||= {
                      upsert: {},
                      delete: {},
-                     assignments: [],
-                     links: [],
+                     assignments: {},
+                     deassignments: {},
```
Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: anyone got a better name for these buffers?  both `unassignments` and `deassignments` and the link equivalents seem a little gross.

**Comment by CGDS** [comment]
Bulk link code added, ready for a round of review

File: lib/policy_machine_storage_adapters/active_record.rb (hunk 6 above)
Comment (DMcKinnon-mdsol) [inline lib/policy_machine_storage_adapters/active_record.rb]: `assignments_to_remove`, `~assignments`

Comment (DMcKinnon-mdsol) [inline lib/policy_machine_storage_adapters/active_record.rb]: Oooo `divorces`

File: lib/policy_machine/policy_element.rb [hunk 7]
```diff
@@ -65,7 +65,9 @@ def unassign(dst_policy_element)
     # Removes self and any assignments to or from self. Does not remove any other elements.
     # Returns true if persisted object was successfully removed.
     def delete
-      if self.stored_pe && self.stored_pe.persisted
+      # If the stored ppe is not persisted, no need to delete it, but if buffering
+      # all bets are off, and the storage adapter needs to handle the situation
+      if pm_storage_adapter.buffering? || (self.stored_pe && self.stored_pe.persisted)
```
Comment (DMcKinnon-mdsol) [inline lib/policy_machine/policy_element.rb]: No `try`?

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 8]
```diff
@@ -376,7 +395,7 @@ def assign(src, dst)
     end
 
     def assign_later(parent:, child:)
-      buffers[:assignments] << [parent, child]
+      buffers[:assignments].merge!([parent.unique_identifier, child.unique_identifier] => [parent, child])
```
Comment (DMcKinnon-mdsol) [inline lib/policy_machine_storage_adapters/active_record.rb]: Are we using `unique_identifier` for assignments and `id` for links? Whoops, my bad.

File: spec/policy_machine_storage_adapters/active_record_spec.rb [hunk 9]
```diff
@@ -107,6 +107,207 @@
...
+
+      describe '#bulk_persist' do
+        let(:pm) { PolicyMachine.new(name: 'AR PM', storage_adapter: PolicyMachineStorageAdapter::ActiveRecord) }
+        let(:user) { pm.create_user('alice') }
+        let(:caffeinated) { pm.create_user_attribute('caffeinated') }
+        let(:decaffeinated) { pm.create_user_attribute('decaffeinated') }
+
+        describe 'policy element behavior' do
+          it 'deletes a policy element that has been created and then deleted in a persistence buffer' do
+            user, attr = pm.bulk_persist do
```
Comment (DMcKinnon-mdsol) [inline spec/policy_machine_storage_adapters/active_record_spec.rb]: Mind finding a different variable name than a bult-in word?

File: spec/policy_machine_storage_adapters/active_record_spec.rb [hunk 10]
```diff
@@ -107,6 +107,207 @@
...
+              user = pm.create_user('alice')
+              attr = pm.create_user_attribute('caffeinated')
+              user.delete
+
+              [user, attr]
+            end
+
+            expect(pm.user_attributes).to eq [attr]
+            expect(pm.users).to be_empty
```
Comment (DMcKinnon-mdsol) [inline spec/policy_machine_storage_adapters/active_record_spec.rb]: Could also do/check `user.persisted?`. Not saying it's better or worse than this way.

File: spec/policy_machine_storage_adapters/active_record_spec.rb [hunk 11]
```diff
@@ -107,6 +107,207 @@
...
+          end
+
+          it 'creates a record if the record is created, deleted and then recreated inside a persistence buffer' do
+            user, attr = pm.bulk_persist do
+              pm.create_user('alice').delete
+              attr = pm.create_user_attribute('caffeinated')
+              user = pm.create_user('alice')
+
+              [user,attr]
```
Comment (DMcKinnon-mdsol) [inline spec/policy_machine_storage_adapters/active_record_spec.rb]: Space after comma.

File: spec/policy_machine_storage_adapters/active_record_spec.rb (hunk 10 above)
Comment (DMcKinnon-mdsol) [inline spec/policy_machine_storage_adapters/active_record_spec.rb]: Actually, what do you think about directly testing if objects are persisted instead?

File: spec/policy_machine_storage_adapters/active_record_spec.rb [hunk 12]
```diff
@@ -107,6 +107,207 @@
...
+
+            expect(user.connected?(caffeinated)).to be true
+            expect(caffeinated.connected?(decaffeinated)).to be false
+          end
+
+          it 'deletes preexisting assignments removed in the persistence buffer' do
+            caffeinated.assign_to(decaffeinated)
+            pm.bulk_persist do
+              user.assign_to(caffeinated)
+              user.assign_to(decaffeinated)
```
Comment (DMcKinnon-mdsol) [inline spec/policy_machine_storage_adapters/active_record_spec.rb]: How is this relevant to this test? Just extra chaos? May as well add an expect for it if you're keeping it.

File: spec/policy_machine_storage_adapters/active_record_spec.rb [hunk 13]
```diff
@@ -107,6 +107,207 @@
...
       end
+
+      describe '#bulk_persist' do
+        let(:pm) { PolicyMachine.new(name: 'AR PM', storage_adapter: PolicyMachineStorageAdapter::ActiveRecord) }
+        let(:user) { pm.create_user('alice') }
+        let(:caffeinated) { pm.create_user_attribute('caffeinated') }
+        let(:decaffeinated) { pm.create_user_attribute('decaffeinated') }
+
+        describe 'policy element behavior' do
+          it 'deletes a policy element that has been created and then deleted in a persistence buffer' do
```
Comment (DMcKinnon-mdsol) [inline spec/policy_machine_storage_adapters/active_record_spec.rb]: Given that all these specs end with `in a persistence buffer`, it might be worth just putting them in a context block.

**Comment by DMcKinnon-mdsol** [comment]
Might be worth also adding some tests for update behavior in buffer.

File: lib/policy_machine/policy_element.rb (hunk 7 above)
Comment (CGDS) [inline lib/policy_machine/policy_element.rb]: There is no try.  What did you want the try on?  the `buffering?`?

File: spec/policy_machine_storage_adapters/active_record_spec.rb (hunk 9 above)
Comment (CGDS) [inline spec/policy_machine_storage_adapters/active_record_spec.rb]: fair point

File: spec/policy_machine_storage_adapters/active_record_spec.rb (hunk 13 above)
Comment (CGDS) [inline spec/policy_machine_storage_adapters/active_record_spec.rb]: I think i can just delete, its already in a `#bulk_persist` describe block

**Comment by HonoreDB** [comment]
LGTM

File: lib/policy_machine_storage_adapters/active_record.rb (hunk 6 above)
Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: ...does this imply you want to rename `assignments` as `marriages`? Or is the choice just influenced by the wedding you just went to...

Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: you can't divorce your children @DMcKinnon-mdsol, or so my parents tell me.

File: lib/policy_machine_storage_adapters/active_record.rb (hunk 8 above)
Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: for the record, we can't use `id` in the create case because the `id`s may or may not exist at the time we insert them into the buffer, depending on whether or not the model has been persisted.

**Comment by CGDS** [comment]
comments addressed

File: lib/policy_machine/policy_element.rb (hunk 7 above)
Comment (PoloPro) [inline lib/policy_machine/policy_element.rb]: I guess the latter clause could be replaced with `self.stored_pe.try(:persisted)`? Is that what you meant, Don?

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 14]
```diff
@@ -184,13 +189,27 @@ def self.bulk_destroy(elements)
...
-      def self.bulk_assign(parents_and_children)
-        id_pairs = parents_and_children.map { |parent, child| [parent.id, child.id]  }
+      def self.bulk_assign(pairs_hash)
+        id_pairs = pairs_hash.values.map { |parent, child| [parent.id, child.id]  }
         Assignment.import([:parent_id, :child_id], id_pairs, on_duplicate_key_ignore: true)
       end
 
-      def self.bulk_link(parents_and_children)
-        id_pairs = parents_and_children.map { |parent, child| [parent.id, child.id, parent.policy_machine_uuid, child.policy_machine_uuid] }
+      def self.bulk_unassign(pairs_hash)
```
Comment (PoloPro) [inline lib/policy_machine_storage_adapters/active_record.rb]: This looks so innocuous now I hardly believe it.

**Comment by PoloPro** [comment]
LGTM. Can merge if you're happy with it, @CGDS
